### PR TITLE
[CI] Add PHP 8.0 to Github actions tests

### PR DESCRIPTION
**Pull request by samsonasik**

### CODE REVIEW

**Comment by samsonasik** [comment]
Got php 8 downgrade errors:

```bash
Downgrade Class On Object To Get Class Rector (Rector\DowngradePhp80\Tests\Rector\ClassConstFetch\DowngradeClassOnObjectToGetClassRector\DowngradeClassOnObjectToGetClassRector)
 ✘  with data set #0
   │
   │ rules/downgrade-php80/tests/Rector/ClassConstFetch/DowngradeClassOnObjectToGetClassRector/Fixture/fixture.php.inc
   │ Failed asserting that string matches format description.
   │ --- Expected
   │ +++ Actual
   │ @@ @@
   │  {
   │      public function run($object)
   │      {
   │ -        return get_class($object);
   │ +        return $object::class;
   │      }
   │  }
   │
   │ /home/runner/work/rector/rector/packages/testing/src/PHPUnit/AbstractRectorTestCase.php:486
   │ /home/runner/work/rector/rector/packages/testing/src/PHPUnit/AbstractRectorTestCase.php:308
   │ /home/runner/work/rector/rector/rules/downgrade-php80/tests/Rector/ClassConstFetch/DowngradeClassOnObjectToGetClassRector/DowngradeClassOnObjectToGetClassRectorTest.php:21
   │


Downgrade Return Static Type Declaration Rector (Rector\DowngradePhp80\Tests\Rector\FunctionLike\DowngradeReturnStaticTypeDeclarationRector\DowngradeReturnStaticTypeDeclarationRector)
 ✘  with data set #0
   │
   │ rules/downgrade-php80/tests/Rector/FunctionLike/DowngradeReturnStaticTypeDeclarationRector/Fixture/fixture.php.inc
   │ /home/runner/work/rector/rector/src/Application/FileProcessor.php:143
   │ /home/runner/work/rector/rector/packages/testing/src/PHPUnit/AbstractRectorTestCase.php:439
   │ /home/runner/work/rector/rector/packages/testing/src/PHPUnit/AbstractRectorTestCase.php:308
   │ /home/runner/work/rector/rector/rules/downgrade-php80/tests/Rector/FunctionLike/DowngradeUnionTypeReturnDeclarationRector/DowngradeUnionTypeReturnDeclarationRectorTest.php:23
   │

Downgrade Union Type Typed Property Rector (Rector\DowngradePhp80\Tests\Rector\Property\DowngradeUnionTypeTypedPropertyRector\DowngradeUnionTypeTypedPropertyRector)
 ✘  with data set #0
   │
   │ Rector\Core\Exception\NotImplementedException: PhpParser\Node\UnionType
   │
   │ /home/runner/work/rector/rector/packages/static-type-mapper/src/Mapper/PhpParserNodeMapper.php:45
   │ /home/runner/work/rector/rector/packages/static-type-mapper/src/StaticTypeMapper.php:82
   │ /home/runner/work/rector/rector/rules/downgrade-php74/src/Rector/Property/AbstractDowngradeTypedPropertyRector.php:45
   │ /home/runner/work/rector/rector/src/Rector/AbstractRector.php:231
   │ /home/runner/work/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:200
   │ /home/runner/work/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:114
   │ /home/runner/work/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:223
   │ /home/runner/work/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:114
   │ /home/runner/work/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:223
   │ /home/runner/work/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:91
   │ /home/runner/work/rector/rector/src/PhpParser/NodeTraverser/RectorNodeTraverser.php:119
   │ /home/runner/work/rector/rector/src/Application/FileProcessor.php:143
   │ /home/runner/work/rector/rector/packages/testing/src/PHPUnit/AbstractRectorTestCase.php:439
   │ /home/runner/work/rector/rector/packages/testing/src/PHPUnit/AbstractRectorTestCase.php:308
   │ /home/runner/work/rector/rector/rules/downgrade-php80/tests/Rector/Property/DowngradeUnionTypeTypedPropertyRector/DowngradeUnionTypeTypedPropertyRectorTest.php:2
```

**Comment by samsonasik** [comment]
@TomasVotruba It seems `isAtLeastPhpVersion()` not works for php 8, debug with the following:

```php
        dump(
            PHP_VERSION,
            $this->isAtLeastPhpVersion(PhpVersionFeature::CLASS_ON_OBJECT)
        );
```

got:

```php
"8.0.0" (5)
false
```

**Comment by samsonasik** [comment]
It seems because of test definition:

```php
    protected function getPhpVersion(): int
    {
        return PhpVersionFeature::CLASS_ON_OBJECT - 1;
    }
```

**Comment by samsonasik** [comment]
Got weird error at `Rector\Naming\Tests\Rector\Assign\RenameVariableToMatchMethodCallReturnTypeRector\Php74Test`

```bash
Php74 (Rector\Naming\Tests\Rector\Assign\RenameVariableToMatchMethodCallReturnTypeRector\Php74)
 ✘  with data set #0
   │
   │ rules/naming/tests/Rector/Assign/RenameVariableToMatchMethodCallReturnTypeRector/FixturePhp74/arrow_function.php.inc
   │ Failed asserting that string matches format description.
   │ --- Expected
   │ +++ Actual
   │ @@ @@
   │      {
   │          $datetime = new DateTime();
   │  
   │ -        $fn1 = fn($x) => $x + $dateTimeZone = $datetime->getTimezone();
   │ +        $fn1 = fn($x) => $x + $timezone = $datetime->getTimezone();
   │      }
   │  }
```

**Comment by samsonasik** [comment]
More error in test in php 8 : 

```bash
To String To Method Call Rector (Rector\MagicDisclosure\Tests\Rector\String_\ToStringToMethodCallRector\ToStringToMethodCallRector)
 ✔  with data set #0
PHP Fatal error:  Cannot use "parent" when current class scope has no parent in /tmp/_temp_fixture_easy_testing/input_d61e4d22d14036671bb8_rename_class_method_and_parent_call.php on line 19
```

seems need to be fixed in somewhere in the code instead

File: rules/downgrade-php80/tests/Rector/FunctionLike/DowngradeReturnStaticTypeDeclarationRector/DowngradeReturnStaticTypeDeclarationRectorTest.php [hunk 1]
```diff
@@ -13,7 +13,7 @@
 final class DowngradeReturnStaticTypeDeclarationRectorTest extends AbstractRectorTestCase
 {
     /**
-     * @requires PHP >= 8.0
+     * @requires PHP >= 9.0
```
Comment (samsonasik) [inline rules/downgrade-php80/tests/Rector/FunctionLike/DowngradeReturnStaticTypeDeclarationRector/DowngradeReturnStaticTypeDeclarationRectorTest.php:16]: @TomasVotruba I set temporary this as test is not working:

```bash
Downgrade Union Type Typed Property Rector (Rector\DowngradePhp80\Tests\Rector\Property\DowngradeUnionTypeTypedPropertyRector\DowngradeUnionTypeTypedPropertyRector)
 ✘  with data set #0
   │
   │ Rector\Core\Exception\NotImplementedException: PhpParser\Node\UnionType
   │
   │ /home/runner/work/rector/rector/packages/static-type-mapper/src/Mapper/PhpParserNodeMapper.php:45
   │ /home/runner/work/rector/rector/packages/static-type-mapper/src/StaticTypeMapper.php:82
   │ /home/runner/work/rector/rector/rules/downgrade-php74/src/Rector/Property/AbstractDowngradeTypedPropertyRector.php:45
   │ /home/runner/work/rector/rector/src/Rector/AbstractRector.php:231
   │ /home/runner/work/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:200
   │ /home/runner/work/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:114
   │ /home/runner/work/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:223
   │ /home/runner/work/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:114
   │ /home/runner/work/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:223
   │ /home/runner/work/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:91
   │ /home/runner/work/rector/rector/src/PhpParser/NodeTraverser/RectorNodeTraverser.php:119
   │ /home/runner/work/rector/rector/src/Application/FileProcessor.php:143
   │ /home/runner/work/rector/rector/packages/testing/src/PHPUnit/AbstractRectorTestCase.php:439
   │ /home/runner/work/rector/rector/packages/testing/src/PHPUnit/AbstractRectorTestCase.php:308
   │ /home/runner/work/rector/rector/rules/downgrade-php80/tests/Rector/Property/DowngradeUnionTypeTypedPropertyRector/DowngradeUnionTypeTypedPropertyRectorTest.php:23
```

Comment (samsonasik) [reply rules/downgrade-php80/tests/Rector/FunctionLike/DowngradeReturnStaticTypeDeclarationRector/DowngradeReturnStaticTypeDeclarationRectorTest.php]: marked to 8.1 seems works, this is temporary to be fixed in the future.

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

File: rules/downgrade-php80/tests/Rector/Property/DowngradeUnionTypeTypedPropertyRector/DowngradeUnionTypeTypedPropertyRectorTest.php [hunk 2]
```diff
@@ -11,7 +11,7 @@
 use Symplify\SmartFileSystem\SmartFileInfo;
 
 /**
- * @requires PHP >= 8.0
+ * @requires PHP = 8.1
```
Comment (TomasVotruba) [inline rules/downgrade-php80/tests/Rector/Property/DowngradeUnionTypeTypedPropertyRector/DowngradeUnionTypeTypedPropertyRectorTest.php:14]: Why this change? It skips the test until 8.1, but it should work on 8.0.
Same for other conditions

**Comment by TomasVotruba** [comment]
> @TomasVotruba It seems `isAtLeastPhpVersion()` not works for php 8, debug with the following:
> 
> ```
>         dump(
>             PHP_VERSION,
>             $this->isAtLeastPhpVersion(PhpVersionFeature::CLASS_ON_OBJECT)
>         );
> ```
> 
> got:
> 
> ```
> "8.0.0" (5)
> false
> ```

Instead of `PHP_VERSION`, it should be `PHP_VERSION_ID`

File: rules/downgrade-php80/tests/Rector/Property/DowngradeUnionTypeTypedPropertyRector/DowngradeUnionTypeTypedPropertyRectorTest.php (hunk 2 above)
Comment (samsonasik) [reply rules/downgrade-php80/tests/Rector/Property/DowngradeUnionTypeTypedPropertyRector/DowngradeUnionTypeTypedPropertyRectorTest.php:14]: The test currently failure, see https://github.com/rectorphp/rector/pull/4779#discussion_r536491712 , this PR is for enabing PHP 8.0 in CI, the fix should be in different PR I think

Comment (TomasVotruba) [reply rules/downgrade-php80/tests/Rector/Property/DowngradeUnionTypeTypedPropertyRector/DowngradeUnionTypeTypedPropertyRectorTest.php:14]: I see, ok then :+1:

**Comment by TomasVotruba** [comment]
Let's go :)

**Comment by TomasVotruba** [comment]
Most of the skipped test cases are fixed here:
https://github.com/rectorphp/rector/pull/4805

**Comment by samsonasik** [comment]
Thank you
